### ISSUE TITLE
[RFC] Stop automatically creating legacy and p2sh-segwit descriptors

### ISSUE THREAD

**Issue by achow101** [bitcoin/bitcoin#23831]
As we see segwit and taproot usage grow, it seems reasonable to stop making wallets that are able to generate legacy and p2sh-segwit addresses. We have already made it difficult to make legacy addresses through the GUI (starting with `-addresstype=legacy` is the only way). The default address type is bech32. So it seems that we should consider stop automatically generating legacy, and probably p2sh-segwit, descriptors.

This would simply be removing the automatic creation of descriptors for the `LEGACY` and `P2SH_SEGWIT` output types for new wallets. We would still retain the ability to spend those outputs, parse those descriptors, and make those types of addresses if the user's wallet has descriptors for these output types. Additionally, if a user wanted to add such a descriptor to their wallet, they could via `importdescriptors`. There can also be some RPC which can add an automatically generated single key descriptor of whatever type that could legacy and p2sh-segwit descriptors to the wallet. Such an RPC would be the generic version of the one that will add `tr` descriptors (or perhaps the same RPC).

**Comment by MarcoFalke** [comment]
I think it would be nice to have an RPC to fill a blank descriptor wallet with specific descriptor types instead of the all-or-blank approach right now.

Haven't thought a lot about not providing legacy/p2sh by default for non-blank wallets, but there are still some places that refuse to upgrade and can only support payout to base58 addresses.

**Comment by Sjors** [comment]
At minimum there should be a way to add them manually, e.g. with #22341 and (the import part of) #23544.

Dropping `LEGACY` by default would be fine with me, if e.g. we do generate them when started with `-addresstype=legacy`. Dropping `P2SH_SEGWIT` seems premature.

But I don't know if there's much added value to dropping 1 out of 4 descriptors. If performance if the main concern, then let's just make the default key pool smaller.

**Comment by Rspigler** [comment]
I agree with removing legacy by default (and argued for that [here](https://github.com/bitcoin-core/gui/pull/459#issuecomment-961328134), but the legacy address option was added back to the GUI in this PR: https://github.com/bitcoin-core/gui/pull/459.

According to https://en.bitcoin.it/wiki/Bech32_adoption almost everyone can now send to Bech32.  Doing this will probably help with the last stragglers.

**Comment by Rspigler** [comment]
I agree with removing legacy by default (and argued for that https://github.com/bitcoin-core/gui/pull/459#issuecomment-961328134, but the legacy address option was added back to the GUI in this PR: https://github.com/bitcoin-core/gui/pull/459.

According to https://en.bitcoin.it/wiki/Bech32_adoption almost everyone can now send to Bech32. Doing this will probably help with the last stragglers.

**Comment by Sjors** [comment]
@Rspigler it's fairly easy to modify the GUI to only show the relevant address options.

**Comment by Xekyo** [comment]
I'd consider it beneficial to further discourage creation of legacy UTXOs with new wallets both because I don't see a point to wasting block weight and it helping us move past third-party malleable transactions. The only reasons I'd see for receiving legacy outputs from a newly created wallet these days would be to match the change type to a recipient's output (which hopefully eventually goes away by others not using legacy addresses), and wanting to be able to sign a message that is clearly associated with a specific UTXO (which hopefully gets resolved by some generic message signing standard).

**Comment by 1440000bytes** [comment]
NACK

We dont't need to be dictators as bitcoin core developers Even if we try to, users will try different type of addresses in transactions. Bitcoin Core Wallet is the not the main wallet used for all transactions.

Example: 

```
Block 754309:

758 transactions
638 witness tx

120 tx with no witness - 73 have different version and locktime
```

Are you trying to police what people do with their transactions when segwit and taproot was a soft soft OPTIN?

This will only affect how core wallet wallet works not bitcoin.

**Comment by achow101** [comment]
> We dont't need to be dictators as bitcoin core developers even if we try to, users will try different type of addresses in transactions. Bitcoin Core Wallet is the not the main wallet used for all transactions.
>
> Are you trying to police what people do with their transactions when segwit and taproot was a soft fork OPTIN?
>
> This will only affect how core wallet works not bitcoin.

I think you misunderstand the purpose of this suggestion. I'm not trying to force anything onto the network or "be dictators". I'm not saying that this will affect users who do not use Bitcoin Core. The purpose is to improve the user experience for users of the wallet, and that includes not storing data that will likely never be used, and also making it harder to have to spend more money unintentionally.

Even so, this suggestion does not mean that legacy addresses will be completely unsupported in the wallet. Anyone who has a legacy address descriptor will still be able to make legacy addresses. If you want legacy addresses and the wallet does not have such a descriptor, one can be imported using `importdescriptors`. After #25907, the process for adding such a descriptor will also become much easier.

**Comment by 1440000bytes** [comment]
> > We dont't need to be dictators as bitcoin core developers even if we try to, users will try different type of addresses in transactions. Bitcoin Core Wallet is the not the main wallet used for all transactions.
> > Are you trying to police what people do with their transactions when segwit and taproot was a soft fork OPTIN?
> > This will only affect how core wallet works not bitcoin.
> 
> I think you misunderstand the purpose of this suggestion. I'm not trying to force anything onto the network or "be dictators". I'm not saying that this will affect users who do not use Bitcoin Core. The purpose is to improve the user experience for users of the wallet, and that includes not storing data that will likely never be used, and also making it harder to have to spend more money unintentionally.
> 
> Even so, this suggestion does not mean that legacy addresses will be completely unsupported in the wallet. Anyone who has a legacy address descriptor will still be able to make legacy addresses. If you want legacy addresses and the wallet does not have such a descriptor, one can be imported using `importdescriptors`. After #25907, the process for adding such a descriptor will also become much easier.

I agree with you and sorry for misunderstanding. Go ahead, I already love PSBT created by you however its not easy and we need to look at privacy, security, ui,ux, etc.

**Comment by luke-jr** [comment]
>wasting block weight

Consuming, but not necessarily wasting (a value judgement). Consuming more weight with less actual cost helps keep block sizes down. Block sizes are already harmfully large, so that is good for Bitcoin.

**Comment by 1440000bytes** [comment]
Even though I agree with @achow101 its okay we allow:

```stop making wallets that are able to generate legacy and p2sh-segwit addresses. We have already made it difficult to make legacy addresses through the GUI (starting with -addresstype=legacy is the only way). The default address type is bech32. So it seems that we should consider stop automatically generating legacy, and probably p2sh-segwit, descriptors.
```

dont't be be so cruel in bitcoin

**Comment by 1440000bytes** [comment]
> > wasting block weight
> 
> Consuming, but not necessarily wasting (a value judgement). Consuming more weight with less actual cost helps keep block sizes down. Block sizes are already harmfully large, so that is good for Bitcoin.

Looking at the demand for block space space in Bitcoin..  I think we should make legacy address as default.

**Comment by Rspigler** [comment]
This doesn't make sense to me. If we want to make better use of block space, we should want to take up less space with each transaction?
https://bitcoinops.org/en/tools/calc-size/

**Comment by luke-jr** [comment]
Segwit/Taproot transactions do not use significantly less space. They just allow blocks to be larger.

**Comment by 1440000bytes** [comment]
> Segwit/Taproot transactions do not use significantly less space. They just allow blocks to be larger.

+1k

**Comment by 1440000bytes** [comment]
Last pull request: https://github.com/bitcoin/bitcoin/pull/24362

People said a few things

I reacted..

I dont't want to share what they said.. I will remember

Can we stop creating anything by default? What's wrong with p2sh-segwit?

**Comment by 1440000bytes** [comment]
What's wrong with users trying different type of addresses

**Comment by Xekyo** [comment]
> Segwit/Taproot transactions do not use significantly less space. They just allow blocks to be larger.

P2TR are 17% smaller than P2PKH. P2TR inputs have a serialized byte length of 107 B vs 147 or 148 B for P2PKH, and while P2TR outputs are 43 B vs 34 B for P2PKH. So in total 150 B vs 181/182 B. The size reduction is bigger for multisig applications that transition to aggregate key constructions.

**Comment by luke-jr** [comment]
Not only is that trivial, it's also "bytes" that don't have an actual cost. In other words, they're just additional weight, not actually larger.

**Comment by Xekyo** [comment]
I would have actually expected you to like P2TR. It has significantly less witness data per non-witness data than prior segwit output types.

Either way, asking for bad defaults in software that cause its users to pay more for equivalent outcomes to achieve a global reduction of the overall throughput doesn't sound like a compelling approach—game theoretically speaking.

**Comment by 1440000bytes** [comment]
bitcoin for users who can can use all type of address based on need not BLOCKSTREAM and CHAINCODELABS telling them what is good/bad

**Comment by 1440000bytes** [comment]
We dont need certificate for address from core developers. Just Let us use everything.

**Comment by aureleoules** [comment]
> Just Let us use everything.

You would.
> We would still retain the ability to spend those outputs, parse those descriptors, and make those types of addresses if the user's wallet has descriptors for these output types. Additionally, if a user wanted to add such a descriptor to their wallet, they could via importdescriptors

**Comment by fanquake** [comment]
> bitcoin for users who can can use all type of address based on need not BLOCKSTREAM and CHAINCODELABS telling them what is good/bad

> We dont need certificate for address from core developers. Just Let us use everything.

> https://github.com/bitcoin/bitcoin/pull/25600#issuecomment-1255332319 (other thread)

@1440000bytes Please keep your commentary on topic, and stop with the random accusations / hyperbole. This repository is not the place for it.

**Comment by 1440000bytes** [comment]
> > bitcoin for users who can can use all type of address based on need not BLOCKSTREAM and CHAINCODELABS telling them what is good/bad
> 
> > We dont need certificate for address from core developers. Just Let us use everything.
> 
> > [#25600 (comment)](https://github.com/bitcoin/bitcoin/pull/25600#issuecomment-1255332319) (other thread)
> 
> @1440000bytes Please keep your commentary on topic, and stop with the random accusations / hyperbole. This repository is not the place for it.

Sorryt. I still think we should have multiple addresses.

**Comment by 1440000bytes** [comment]
I would love to see usage for new addresses likes P2TR but I don't wan't want to remove options for all old types in descriptor walllet

**Comment by Rspigler** [comment]
@1440000bytes This PR wouldn't effect such a scenario because (emphasis mine).

> This would simply be removing the automatic creation of descriptors for the LEGACY and P2SH_SEGWIT output types for /new wallets/. We would still retain the ability to spend those outputs, parse those descriptors, and make those types of addresses if the user's wallet has descriptors for these output types.

**Comment by 1440000bytes** [comment]
> @1440000bytes This PR wouldn't effect such a scenario because (emphasis mine).
> 
> > This would simply be removing the automatic creation of descriptors for the LEGACY and P2SH_SEGWIT output types for /new wallets/. We would still retain the ability to spend those outputs, parse those descriptors, and make those types of addresses if the user's wallet has descriptors for these output types.

_This would simply be removing the automatic creation of descriptors for the LEGACY and P2SH_SEGWIT_

This is not required. This is what I don't like cares issues so I am commenting. And there is nothing wrong with allowing users to create wallets with all types of descriptors. I do not agree with the opinion that this will hep in efficient block usage. Bitcoin is not an efficient financial network, its censorship resistant network where users should be allowed to do what they want. They could still import descriptors but default matters and affects the way people use core right now.

**Comment by Sjors** [comment]
Another thought: we could lazy-generate the descriptors the first time a user calls `getnewaddress`, creates a new address in the GUI, etc.

**Comment by murchandamus** [comment]
Since even Coinbase now seems to be working on sending to bech32m (Coinbase Custody already being able and Coinbase retail app supposedly getting it soon), we are even getting into the territory where we could default to P2TR addresses eventually. Surely that means that we could stop creating legacy and wrapped segwit address descriptors on new wallets by default.

**Comment by murchandamus** [comment]
We could have probably done this last year, but we definitely should do it this year.
